### ISSUE TITLE
[ENH] enhancements to `neuralforecast` adapter

### ISSUE THREAD

**Issue by yarnabrina** [sktime/sktime#6006]
I've mentioned few limitations in the first draft of the adapter added in #5962. This issue is to discuss and track these enhancements.

- [ ] support of static (constant over time) exogenous features
    - how useful?
    - any predictive capability for separate model per series?
- [ ] support of historical (unknown for prediction horizon) exogenous features
    - alternative to `ForecastX`?
    - compatiblity with `sktime`?
- [ ] support of panel/hierarchical datasets
    - any existing example in sktime?
- [ ] support of probabilistic forecasting
    - depends on loss function?
- [ ] support of in-sample predictions
- [ ] support of fitted parameters
    - expose best parameters from `Auto*` models?

This list can be edited to add other extension ideas. This list intentionally skips addition of other models in `neuralforecast`, as those are not adapter enhancements. But that may be a good first issue for new contributors, and can be very useful to users (e.g. I'd love to use the `Auto*` versions).

FYI @sktime/core-developers @jmoralez

**Comment by fkiraly** [comment]
> But that may be a good first issue for new contributors

Ok, adding the "good first issue" label then.

**Comment by fkiraly** [comment]
> support of static (constant over time) exogenous features 

This is "marginal" in the sense that there is not a dedicated representation of such features in *any* of the about 20 time series data types known by `sktime`.

However, I am entusiastic about adding `temporian` `EventSet`, which afaik is the first (non-hacky, well-designed) representation that allows values at different index depth.

FYI @ianspektor - perhaps `neuralforecast` could be the first instance of a forecaster that supports time-constant features which we test after adding `EventSet` as a data type?

**Comment by ianspektor** [comment]
Hey 👋🏼 The `EventSet` doesn't support static features, maybe there was a misunderstanding in one of our calls. What the `EventSet` would enable is having data stored for and perform operations on different index levels (e.g. store_id and store_id+product_id, I think this is what you refer to as different index depths) and easily combine those into a single output, but the data inside an EventSet is always attached to a timestamp. There's straightforward ways of merging/broadcasting an array of static features into an `EventSet` though, which would assign the correct value to each event.

cc. @achoum

**Comment by fkiraly** [comment]
@ianspektor, @achoum, thanks for explaining - in that case, it is a bit unfortunate and I am starting to wonder why.

It seems like you support values at different index depths in all possible combinations *except* for time being the index. That strikes me as a strange - and implementation-wise, in-principle unnecessary? - limitation.

Because, isn't time just another index level? I don't see an a-priori reason why it should be treated separately. Though possibly there are good design reasons that would become only apparent to me if I look more closely at `EventSet`. Explanations appreciated.

**Comment by fkiraly** [comment]
> having data stored for and perform operations on different index levels (e.g. store_id and store_id+product_id, I think this is what you refer to as different index depths) 

yes - not sure if there is commonly used terminology here. I imagine multi-indices with an abstract data type that has indices as a tree (or ordered sub-set), and values can be stored at leaves (common) or nodes (uncommon), with depth being used as usually when discussing tree structures.

**Comment by yarnabrina** [comment]
> > But that may be a good first issue for new contributors
> 
> Ok, adding the "good first issue" label then.

Deviation from the above discussion of static features: I meant good first issue for interfacing new models which I intentionally didn't add in this issue tasks and kept it focused for adapter only. I felt adapter generalisation may be next level. @fkiraly please edit the task list if you want to track interfacing part in this issue, or may be we can do in a new issue?

---

Another question for @fkiraly is about generalising support for future-unknown features and forecastsers with direct support of panel data. Do we have any example in sktime for these? I believe neuralforecast models support these, so how do we support? Is there any sample class/code you can share that I can use as a reference?

---

Regarding constant features, I was mainly thinking if they are useful or not. Definitely not in current case of model per series, but may be if we add panel support. But it seems you are discussing about container support? I don't follow the problem. These can be a column with constant values, so shouldn't all types support it? Of course it'll be wasting a lots of storage unnecessarily, but any other technical issues?

Really sorry for asking very naive/obvious questions, but it's not clear to me.

**Comment by fkiraly** [comment]
> I meant good first issue for interfacing new models

Should we remove the "good first issue" tag then?

> support for future-unknown features and forecastsers with direct support of panel data

As far as I know:

* a small but significant minority of forecasters supports panel data, e.g., reduction forecasters; further non-trivial panel data logic can be added through reconciliation or composition (e.g., hierarchical ensemble, tuning across all instances)
* as far as I know, no forecaster supports future-unknown variables natively. This support can be added via `ForecastX` - that's also what the `darts` one seems to do internally?
